### ISSUE TITLE
How to constrain features by complicated geo constraints

### ISSUE THREAD

**Issue by grtjn** [marklogic-community/marklogic-geo-data-services#48]
Not sure if it is possible or not yet, so may end up as a feature request. I'd like to get a subset of features back, filtered by a potentially complicated and/or query of polygons/circles. Someone might have drawn 3 circles, boxes or polygons on a map, and want to see the matching features in either the intersect of those 3, or maybe the union of those 3.

**Comment by jkerr5** [comment]
The current query capabilities are defined by the Esri feature service API:  https://developers.arcgis.com/rest/services-reference/query-feature-service-layer-.htm so those are the types of geometries that can be specified as of now. Those are coupled with the "bounding query" as specified via the layer configuration in the service descriptors.

I think we need to make this into an enhancement request and have more discussion on what kind of clients you want to be able to use the GDS API and how.

**Comment by grtjn** [comment]
This might be covered by #47 already. I am really talking about constraints sent in along with the request, so configs won't do. I did understand you can specify a bounding box, but that is far too limited. You'd want to be able to do and's and or's of facet constraints, arbitrary geo constraints, and full-text constraints, essentially anything you'd be able to do with /v1/search.

Regarding kinds of clients: grove apps, or any other kind of custom built front-end webapp (typically via some middle-tier). Regarding use cases: as described above..

**Comment by grtjn** [comment]
As shown separately, I managed to do this by passing in a structured query as query.where, and providing a query.options with search option name as well. The function that handles the query.where has extra logic to detect if query.where contains an object, and if so, assumes a structured query. If not, it does as before. It could be extended to accept a cts query as well, or maybe even a search string (qtext-style) that would be parsed with search:parse..

**Comment by grtjn** [comment]
As mentioned above I managed to do this by making a relatively small, non-breaking change to geoQueryService. It allows you to send in the following request:

```
POST /v1/resources/geoQueryService
{
  params: {
    method: 'query',
    id: 'my-service-name',
    layer: 0 /* layer id */,
    maxRecordCount: 5000
  },
  query: {
    where: { search: { query: /* your structuredQuery */ } }
    optionsName: 'all'
    returnGeometry: true
  }
}
```

I am working on providing my changes to geoQueryService as a PR for review, but need to synch with 1.1.1 release before I can do so. My project is still running 1.1.0.

**Comment by mfgumban** [comment]
Included in 1.2.0 release
